### DISCUSSION TITLE
Can support TileMap in 3D?

### DISCUSSION THREAD

**Discussion by colorworlds** [godotengine/godot-proposals#4130] [3D]
GridMap cannot replace TileMap in 3D, if I want to make a game like **"Don't Starve"**.   

I tried it.  
If two GridMap are stacked in one layer, the rendering order will change, when you rotate the camera and look at it from the back.  
I think GridMap and TileMap are different things
TileMap doesn't have a 3D version, such as TileMap3D, but Sprite Node have a 3D version.

**Comment by theraot** [comment]
Yes, they are different things… Yet, I don't understand what you are trying to do. Why are you stacking two GridMaps? and how supporting that would make it more like TileMap?

---

Since you mention "Don't  Starve", I believe what you want is to have a TileMap in a 3D plane. I can think of three approaches:

- Have a GridMap of quad tiles. This will also give you navigation and collisions.
- Render a TileMap to a Viewport, and put the Viewport texture on the 3D plane. For navigation and collisions, you could have other invisible Node2D on the same space as the TileMap and let them use 2D navigation and collisions… And then have 3D objects that represent them. It is not perfect, but it is something.
- A custom tile shader (and the work that comes with it). And for navigation and collisions, perhaps you can work with physics and navigation servers?

Edit: to be clear, I would also like an specific solution for this. Yet, the two GridMaps thing is confusing me to the point I'm not sure we are talking about the same thing.

---

Anyway, I don't see how two GridMaps would help. If you want multiple layers at different vertical positions in 3D, you make them in the same GridMap.

**Reply by colorworlds** [reply]
I made a game project about 9 months ago.   
At that time, I encountered the problem of disordered rendering order when using GridMap.  
I didn't realize it was a bug before. I thought it was determined by the characteristics of GridMap. 

I made a page to describe the problem in detail:  [render bug](https://gitee.com/jumpcoder/godot-3.2-3d-randerbug) 



This rendering bug has been fixed in 3.3. 

However, the GridMap is still difficult to replace the TileMap, because many functions of GridMap are difficult to use.   

For example, GridMap lack autotile function now. At the same time, The processing logic of TileMap and GridMap is different.  


In addition, the difference between TileMap and GridMap is emphasized because I see some suggestions to rename TileMap to GridMap3D.  

I think if this is done, it will become difficult to use Godot to make some game types.

**Reply by theraot** [reply]
I see, I was unaware that was a bug. It had not occurred to me to use GridMaps like that anyway.

---

GridMap needs usability improvement, we can agree on that.

I'm guessing both 2D autotiling like TileMap and also 3D autotiling are useful. Autotile on 3D would be another dimension of complexity, literally. I can imagine the logic, but I'm having a hard time imagining the UI. And, that would be useful for 3D terrain by the way. With overhangs, bridges, and caves.

With this idea and https://github.com/godotengine/godot-proposals/discussions/4079 I'm beginning to think we would benefit from a lower level tool that uses a MeshLibrary more directly and then we can build GridMap and other similar solution on top of it.

---

Renaming TileMap to TileMap2D would be OK, in fact that would ease introducing a TileMap3D. Similarly renaming GridMap to GridMap3D would be OK.

Renaming TileMap to GridMap3D does not make sense to me.

The renames could make some things harder to figure out, but it would not cause making games to be harder overall.
